torch/nn/modules/_functions.py

```diff
@@ -80,45 +97,73 @@ def backward(self, grad_output):
-            sum_dy, sum_dy_xmu = torch.split(combined, num_channels)
-
-            # backward pass for gradient calculation
-            grad_input = torch.batch_norm_backward_elemt(
+        if saved_input.numel() > 0:
+            # calculate local stats as well as grad_weight / grad_bias
+            sum_dy, sum_dy_xmu, grad_weight, grad_bias = torch.batch_norm_backward_reduce(
                 grad_output,
                 saved_input,
                 mean,
                 invstd,
                 weight,
-                sum_dy,
-                sum_dy_xmu,
-                count_tensor
+                self.needs_input_grad[0],
+                self.needs_input_grad[1],
+                self.needs_input_grad[2]
             )
 
-        # synchronizing of grad_weight / grad_bias is not needed as distributed
-        # training would handle all reduce.
-        if weight is None or not self.needs_input_grad[1]:
-            grad_weight = None
-
-        if weight is None or not self.needs_input_grad[2]:
-            grad_bias = None
+            if self.needs_input_grad[0]:
+                # synchronizing stats used to calculate input gradient.
+                num_channels = sum_dy.shape[0]
+                combined = torch.cat([sum_dy, sum_dy_xmu], dim=0)
+                torch.distributed.all_reduce(
+                    combined, torch.distributed.ReduceOp.SUM, process_group, async_op=False)
+                sum_dy, sum_dy_xmu = torch.split(combined, num_channels)
+
+                # backward pass for gradient calculation
+                grad_input = torch.batch_norm_backward_elemt(
+                    grad_output,
+                    saved_input,
+                    mean,
+                    invstd,
+                    weight,
+                    sum_dy,
+                    sum_dy_xmu,
+                    count_tensor
+                )
+            # synchronizing of grad_weight / grad_bias is not needed as distributed
+            # training would handle all reduce.
+            if weight is None or not self.needs_input_grad[1]:
+                grad_weight = None
+
+            if weight is None or not self.needs_input_grad[2]:
+                grad_bias = None
+        else:
+            num_channels = saved_input.shape[1]
+            if self.needs_input_grad[0]:
+                # launch all_reduce to unblock other peer processes
+                combined = torch.zeros(
+                    2 * num_channels,
+                    dtype=saved_input.dtype,
+                    device=saved_input.device
+                )
+                torch.distributed.all_reduce(
+                    combined, torch.distributed.ReduceOp.SUM, process_group, async_op=False)
+
+                # input is empty, we can skip batch_norm_backward_elemt
+                grad_input = torch.zeros_like(saved_input)
+            else:
+                grad_input = None
+
+            grad_weight = torch.zeros(
```

Comment:
This is not needed. If the result will be just zeros, you can just return None which is interpreted by the autograd as a Tensor full of zeros.

Reply:
Does this also apply to `grad_input`. If I leave that as None, is it guaranteed that autograd engine will also interpret that as a tensor of zeros?